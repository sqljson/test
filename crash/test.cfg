TEMPLATE='{
  "%s":{
    "value": %s,
    "array": %s,
    "where": [{ "code": "value", "value": %s },
              { "code": "array", "array": %s },
              { "code": "where",
                "where": [ {"code": "value", "value": %s},
                           {"code": "array", "array": %s} ] } ] },
  "noise": "Some value"
}'


number=(integer unsignedInt positiveInt Quantity Age Distance SimpleQuantity Duration Count Money)

integer_value=40
integer_array="[39, 40, 38]"
unsignedInt_value="$integer_value"
unsignedInt_array="$integer_array"
positiveInt_value="$integer_value"
positiveInt_array="$integer_array"

Quantity_value='{
	"value": 40,
	"comparator": ">", "unit": "sm", "system": "C", "code": "sm"
}'
Quantity_array='[
{
	"value": 30,
	"comparator": ">", "unit": "sm", "system": "C", "code": "sm"
},{
	"value": 40,
	"comparator": ">", "unit": "sm", "system": "C", "code": "sm"
},{
	"value": 20,
	"comparator": ">", "unit": "sm", "system": "C", "code": "sm"
} ]'
Age_value="$Quantity_value"
Age_array="$Quantity_array"
Distance_value="$Quantity_value"
Distance_array="$Quantity_array"
SimpleQuantity_value="$Quantity_value"
SimpleQuantity_array="$Quantity_array"
Duration_value="$Quantity_value"
Duration_array="$Quantity_array"
Count_value="$Quantity_value"
Count_array="$Quantity_array"
Money_value="$Quantity_value"
Money_array="$Quantity_array"


date=(date dateTime instant time Period Timing)

date_value='"1991-12"'
date_array='["1985-05", "1991-12", "1989-05"]'

dateTime_value='"1991-12-31"'
dateTime_array='["1981-11-12", "1991-12-31", "1990-12-31"]'

instant_value='"2002-05-20T09:00:00"'
instant_array='["2002-02-10T09:00:00", "2002-05-20T09:00:00", "2002-01-30T09:00:00"]'

time_value='"11:00:00"'
time_array='["09:00:00", "11:00:00", "08:00:00"]'

Period_value='{"start": "1991-01-01", "end": "1991-12-31"}'
Period_array='[{"start": "1991-01-01", "end": "1991-12-31"},
							 {"start": "1990-01-01", "end": "1991-12-31"},
							 {"start": "1981-01-01"}]'

Timing_value='{"event": ["1991-01-01", "1992-01-01"]}'
Timing_array='[{"event": ["1991-01-01"]},
							 {"event": ["2000-01-01", "1990-01-01"]},
							 {"event": ["1992-01-01"]}]'

string=(string code id markdown HumanName Address)

string_value='"some STring"'
string_array='["some STring", "some another string", "too string ouuu ehh"]'
code_value='"male"'
code_array='["male", "female", "unknown"]'
id_value='"db042450-efa6-11e6-bc64-92361f002671"'
id_array='["e0349ebe-efa6-11e6-bc64-92361f002671",
					 "e034a24c-efa6-11e6-bc64-92361f002671",
				   "e034a364-efa6-11e6-bc64-92361f002671"]'
markdown_value='"\n ## Hello header\n \n __some bold__ or _italic_\n \n "'
markdown_array='["\n # Hello header 1\n \n __some bold__ or _italic_\n \n ",
								 "\n \n ## Hello header 2\n \n __some bold__ or _italic_\n \n \n ",
								 "\n \n ### Hello header 3\n \n __some bold__ or _italic_\n \n "]'

HumanName_value='{"use": "official",
									"text": "NameText",
									"period":  {"start": "1991-01-01", "end": "1991-12-31"},
									"family": ["family 1", "family 2"],
									"given": ["given 1", "given 2"],
									"prefix": ["prefix 1", "prefix 2"],
									"suffix": ["suffix 1", "suffix 2"]}'
HumanName_array="[$HumanName_value, $HumanName_value, $HumanName_value]"

Address_value='{"use": "home", "type":"physical",
								"text": "Some address text", "line": ["line 1", "line 2"],
								"city": "SPB", "District": "DistrictFive",
								"state": "state", "postalCode": "Postal code",
								"period":  {"start": "1991-01-01", "end": "1991-12-31"},
								"country": "Russia" }'
Address_array="[$Address_value, $Address_value, $Address_value]"



token=(boolean code string Identifier Coding CodeableConcept ContactPoint)

boolean_value='"true"'
boolean_array='["true", "false", "true"]'
code_value='"Some code"'
code_array='["some code", "some another code"]'
Identifier_value='{
	"use": "use",
  "type": {"coding": [{"system": "type_coding_system", "code": "type_coding_code"}], "text": "type_text"},
  "system": "system",
  "value": "value",
  "period": {"start": "1991-01-01", "end": "1991-12-31"}
}'
Identifier_array="[$Identifier_value, $Identifier_value]"

Coding_value='{"system": "type_coding_system", "code": "type_coding_code"}'
Coding_array='[{"system": "system_1", "code": "code_1"}, {"system": "system_2", "code": "code_2"}]'
CodeableConcept_value='{
	"coding": [
		{"system": "system_1", "code": "code_1"},
		{"system": "system_2", "code": "code_2"}],
	"text": "text" }'

CodeableConcept_array="[$CodeableConcept_value, $CodeableConcept_value]"
ContactPoint_value='{ "system": "system", "value": "value", "use": "home", "rank": 1 }'
ContactPoint_array="[$ContactPoint_value, $ContactPoint_value]"


reference=(uri Reference)
uri_value='"http://some.url"'
uri_array='["http://datypic.com", "mailto:info@datypic.com","../%C3%A9dition.html","http://datypic.com/prod.html#shirt","urn:example:org","../Ã©dition.html"]'

Reference_value='{"reference": "http://some.url", "display":"display"}'
Reference_array='[
{"reference": "http://datypic.com", "display":"display"}, {"reference": "mailto:info@datypic.com", "display":"display"},
{"reference": "../%C3%A9dition.html", "display": "display"},
{"reference": "http://datypic.com/prod.html#shirt", "display": "display"},
{"reference": "urn:example:org", "display": "display"}]'

uri=(uri oid)
oid_value=$uri_value
oid_array=$uri_array

quantity=(Quantity)





