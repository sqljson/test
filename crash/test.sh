#!/bin/bash
set -e

source ./test.cfg

function mk_query() {
	local SEARCH_TYPE=$1
  local PTH=$2
  local	DATA_TYPE=$3
	local MODIFICATION=""

	if [ "$SEARCH_TYPE" == number ] || [ "$SEARCH_TYPE" == date ]
  then
    MODIFICATION=", 'max'"
  fi

	echo "DO language plpgsql \$\$ BEGIN RAISE info '$SEARCH_TYPE $PTH $DATA_TYPE'; END \$\$;"
  echo "SELECT fhirpath_as_$SEARCH_TYPE(:resource, '$PTH', '$DATA_TYPE' $MODIFICATION);"
}

function gen_pths() {
  local	data_type=$1
  local paths=()
  paths+=(".$data_type.value")
  paths+=(".$data_type.array")
  paths+=(".$data_type.where.where(code=value).value")
  paths+=(".$data_type.where.where(code=array).array")
  paths+=(".$data_type.where.where(code=where).where.where(code=value).value")
  paths+=(".$data_type.where.where(code=where).where.where(code=array).array")
	echo "${paths[@]}"
}

SEARCH_TYPES=(number date string token reference uri  quantity)


echo "drop extension fhirpath;"
echo "create extension fhirpath;"
for search_type in "${SEARCH_TYPES[@]}"
do
  eval DATA_TYPES=( \${$search_type[@]} ) ;
	for data_type in "${DATA_TYPES[@]}"
	do
		_value=$data_type"_value"
		_array=$data_type"_array"
		value=""
		array=""
		eval "value=\$$_value"
		eval "array=\$$_array"
		if [ -n "$value" ]; then
			RESOURCE="$(printf  "$TEMPLATE" "$data_type" "$value" "$array" "$value" "$array" "$value" "$array")"
			echo "select '''$RESOURCE''' resource \\gset"
			paths=($(gen_pths $data_type))
			for path in "${paths[@]}"
			do
				mk_query $search_type $path $data_type
			done
		fi
	done
done


  #- psql -c 'create database root' postgres
  #- cd crash && ./test.sh > test.sql && psql -v ON_ERROR_STOP=1 root < test.sql && cd ..
  #- psql -c 'drop database root' postgres
