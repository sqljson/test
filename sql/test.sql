create extension jsonknife;

SELECT knife_extract('{"a":{"b": {"c": 5.3}}}', '[["a","b","c"]]'); 

SELECT knife_extract('{"a":[{"b": [{"c": 5.3}, {"c": 6.3}]}, {"b": [{"c": 5.3}, {"c": 6.3}]}]}', '[["a","b","c"]]'); 

SELECT knife_extract('{"a": [1, 2, {"b": {"c": 5.3}}, {"b": {"c": 100}}]}', '[["a", 2 ,"b", "c"]]'); 

SELECT knife_extract('{"name": [{"use": "official", "given": ["a", "b"]}, {"use": "common", "given": ["c", "d"]}]}', '[["name",{"use": "official"},"given"]]');
SELECT knife_extract('{"name": [{"use": "official", "given": ["a", "b"]}, {"use": "official", "given": ["c", "d"]}]}', '[["name",{"use": "official"},"given"]]'); 

SELECT knife_extract('{"name": [{"a": {"b": 1}, "c": "ok"}, {"a": {"b": 2}, "c": "fail"}]}', '[["name", {"a":{"b":1}}, "c"]]');

SELECT knife_extract('{"a":{"b": {"c": 5.3, "d": 6.7}}}', '[["a","b","c"],["a","b","d"]]'); 
