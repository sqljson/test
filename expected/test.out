create extension jsonknife;
SELECT knife_extract('{"a":{"b": {"c": 5.3}}}', '[["a","b","c"]]'); 
 knife_extract 
---------------
 {5.3}
(1 row)

SELECT knife_extract('{"a":[{"b": [{"c": 5.3}, {"c": 6.3}]}, {"b": [{"c": 5.3}, {"c": 6.3}]}]}', '[["a","b","c"]]'); 
   knife_extract   
-------------------
 {5.3,6.3,5.3,6.3}
(1 row)

SELECT knife_extract('{"a": [1, 2, {"b": {"c": 5.3}}, {"b": {"c": 100}}]}', '[["a", 2 ,"b", "c"]]'); 
 knife_extract 
---------------
 {5.3}
(1 row)

SELECT knife_extract('{"name": [{"use": "official", "given": ["a", "b"]}, {"use": "common", "given": ["c", "d"]}]}', '[["name",{"use": "official"},"given"]]');
   knife_extract   
-------------------
 {"\"a\"","\"b\""}
(1 row)

SELECT knife_extract('{"name": [{"use": "official", "given": ["a", "b"]}, {"use": "official", "given": ["c", "d"]}]}', '[["name",{"use": "official"},"given"]]'); 
           knife_extract           
-----------------------------------
 {"\"a\"","\"b\"","\"c\"","\"d\""}
(1 row)

SELECT knife_extract('{"name": [{"a": {"b": 1}, "c": "ok"}, {"a": {"b": 2}, "c": "fail"}]}', '[["name", {"a":{"b":1}}, "c"]]');
 knife_extract 
---------------
 {"\"ok\""}
(1 row)

SELECT knife_extract('{"a":{"b": {"c": 5.3, "d": 6.7}}}', '[["a","b","c"],["a","b","d"]]'); 
 knife_extract 
---------------
 {5.3,6.7}
(1 row)

SELECT knife_extract_text(
  '{"a": {"b": [{"c": "l", "d": "o"}, {"c": 1, "d": ["b", "o", "k"]}]}}',
  '[["a","b","c"],["a","b","d"]]'
  );
 knife_extract_text 
--------------------
 {l,o,b,o,k}
(1 row)

SELECT
'^' ||
array_to_string(
  knife_extract_text(
    '{"a": {"b": [{"c": "l", "d": "o"}, {"c": 1, "d": ["b", "o", "k"]}]}}',
    '[["a","b","c"],["a","b","d"]]'
  ),
  '$^'
)
|| '$'
;
    ?column?     
-----------------
 ^l$^o$^b$^o$^k$
(1 row)

SELECT knife_extract_numeric(
'{"a": 1, "b": [2,3], "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);
 knife_extract_numeric 
-----------------------
 {1,2,3,5}
(1 row)

SELECT knife_extract_max_numeric(
'{"a": 1, "b": 2, "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);
 knife_extract_max_numeric 
---------------------------
                         5
(1 row)

SELECT knife_extract_min_numeric(
'{"a": 1, "b": 2, "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);
 knife_extract_min_numeric 
---------------------------
                         1
(1 row)

